File: src/ts-engine/compress.ts
/** C_t append-then-pool — port of compress.py (Float32Array, no numpy). */

export const DEFAULT_K_MAX = 64;
export const DEFAULT_D = 256;
export const DEFAULT_EMA = 0.5;

export type FloatMatrix = Float32Array[];

export function resolveEma(defaultEma = DEFAULT_EMA): number {
  const raw = (process.env.CHAT_COMPRESSOR_EMA || "").trim();
  if (!raw) return defaultEma;
  const v = Number.parseFloat(raw);
  if (!Number.isFinite(v)) return defaultEma;
  return Math.max(0, Math.min(1, v));
}

export function l2Normalize(rows: FloatMatrix, eps = 1e-8): FloatMatrix {
  return rows.map((row) => {
    let n = 0;
    for (let i = 0; i < row.length; i++) n += row[i]! * row[i]!;
    n = Math.sqrt(n);
    const denom = Math.max(n, eps);
    const out = new Float32Array(row.length);
    for (let i = 0; i < row.length; i++) out[i] = row[i]! / denom;
    return out;
  });
}

function cosineAdjacent(a: Float32Array, b: Float32Array): number {
  let na = 0;
  let nb = 0;
  let dot = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    na += a[i]! * a[i]!;
    nb += b[i]! * b[i]!;
    dot += a[i]! * b[i]!;
  }
  na = Math.sqrt(na);
  nb = Math.sqrt(nb);
  if (na < 1e-12 || nb < 1e-12) return -1.0;
  return dot / (na * nb);
}

function pickEvictionIndex(
  protectedRows: boolean[],
  tags: Set<string>[],
): number | null {
  for (let i = 0; i < protectedRows.length; i++) {
    if (!protectedRows[i]) return i;
  }
  for (let i = 0; i < tags.length; i++) {
    if (!tags[i]!.has("identifier")) return i;
  }
  return protectedRows.length ? 0 : null;
}

export function appendThenPool(
  prevC: FloatMatrix | null,
  newRows: FloatMatrix | Float32Array,
  opts: {
    kMax?: number;
    ema?: number;
    protectMask?: boolean[] | null;
    protectKindsTags?: Set<string>[] | null;
  } = {},
): FloatMatrix {
  const kMax = opts.kMax ?? DEFAULT_K_MAX;
  const ema = opts.ema ?? DEFAULT_EMA;
  let rows: FloatMatrix;
  if (Array.isArray(newRows)) {
    rows = newRows.map((r) => (r instanceof Float32Array ? r : Float32Array.from(r)));
  } else {
    rows = [newRows];
  }
  let stacked: FloatMatrix;
  let nPrev = 0;
  if (!prevC || prevC.length === 0) {
    stacked = rows.map((r) => new Float32Array(r));
  } else {
    nPrev = prevC.length;
    stacked = [...prevC.map((r) => new Float32Array(r)), ...rows.map((r) => new Float32Array(r))];
  }

  let n = stacked.length;
  let protectedRows: boolean[];
  let tags: Set<string>[];
  if (!opts.protectMask) {
    protectedRows = Array(n).fill(false);
    tags = Array.from({ length: n }, () => new Set<string>());
  } else {
    const mask = opts.protectMask;
    if (mask.length === n) {
      protectedRows = mask.map(Boolean);
    } else if (mask.length === n - nPrev) {
      protectedRows = [...Array(nPrev).fill(false), ...mask.map(Boolean)];
    } else {
      throw new Error(`protect_mask length ${mask.length} != stacked rows ${n}`);
    }
    if (!opts.protectKindsTags) {
      tags = protectedRows.map((p) => (p ? new Set(["protected"]) : new Set<string>()));
    } else if (opts.protectKindsTags.length === n) {
      tags = opts.protectKindsTags.map((t) => new Set(t));
    } else if (opts.protectKindsTags.length === n - nPrev) {
      tags = [
        ...Array.from({ length: nPrev }, () => new Set<string>()),
        ...opts.protectKindsTags.map((t) => new Set(t)),
      ];
    } else {
      tags = protectedRows.map((p) => (p ? new Set(["protected"]) : new Set<string>()));
    }
  }

  while (stacked.length > kMax) {
    n = stacked.length;
    let bestI = -1;
    let bestSim = -2.0;
    for (let i = 0; i < n - 1; i++) {
      if (protectedRows[i] || protectedRows[i + 1]) continue;
      const sim = cosineAdjacent(stacked[i]!, stacked[i + 1]!);
      if (sim > bestSim) {
        bestSim = sim;
        bestI = i;
      }
    }
    if (bestI >= 0) {
      const merged = new Float32Array(stacked[bestI]!.length);
      for (let j = 0; j < merged.length; j++) {
        merged[j] = ema * stacked[bestI]![j]! + (1.0 - ema) * stacked[bestI + 1]![j]!;
      }
      stacked = [...stacked.slice(0, bestI), merged, ...stacked.slice(bestI + 2)];
      protectedRows = [
        ...protectedRows.slice(0, bestI),
        false,
        ...protectedRows.slice(bestI + 2),
      ];
      tags = [...tags.slice(0, bestI), new Set(), ...tags.slice(bestI + 2)];
      continue;
    }
    let evict = pickEvictionIndex(protectedRows, tags);
    if (evict === null) evict = 0;
    stacked = [...stacked.slice(0, evict), ...stacked.slice(evict + 1)];
    protectedRows = [...protectedRows.slice(0, evict), ...protectedRows.slice(evict + 1)];
    tags = [...tags.slice(0, evict), ...tags.slice(evict + 1)];
  }

  return l2Normalize(stacked);
}

export function liveMask(k: number, kMax?: number): Float32Array {
  if (kMax === undefined || kMax === k) {
    return Float32Array.from({ length: k }, () => 1);
  }
  const mask = new Float32Array(kMax);
  for (let i = 0; i < k; i++) mask[i] = 1;
  return mask;
}

File: engine/src/chat_compressor/compress.py
"""Compression taxonomy and C_t merge (append-then-pool / slot-wise EMA)."""

from __future__ import annotations

import logging
import os
from typing import Literal

import numpy as np

Method = Literal[
    "digest",
    "hot_set",
    "p1",
    "p2",
    "extractive+hot",
    "extractive",
    "p1-debug",
    "query-pack",
]
DEFAULT_K_MAX = 64
DEFAULT_D = 256
# OpenClaw recall-0.5 default (Cursor upstream was 0.7; use env for cursor-parity)
DEFAULT_EMA = 0.5

log = logging.getLogger("chat_compressor.compress")


def resolve_ema(default: float = DEFAULT_EMA) -> float:
    raw = os.environ.get("CHAT_COMPRESSOR_EMA", "").strip()
    if not raw:
        return float(default)
    try:
        return max(0.0, min(1.0, float(raw)))
    except ValueError:
        return float(default)


def l2_normalize(rows: np.ndarray, eps: float = 1e-8) -> np.ndarray:
    norms = np.linalg.norm(rows, axis=-1, keepdims=True)
    return rows / np.maximum(norms, eps)


def _cosine_adjacent(a: np.ndarray, b: np.ndarray) -> float:
    na = float(np.linalg.norm(a))
    nb = float(np.linalg.norm(b))
    if na < 1e-12 or nb < 1e-12:
        return -1.0
    return float(np.dot(a, b) / (na * nb))


def classify_method(*, producer: str, sampled_via: str) -> str:
    if sampled_via == "p2":
        return "p2"
    if sampled_via in {"extractive+hot", "extractive", "hot_set", "p1-debug", "query-pack"}:
        return sampled_via
    if sampled_via == "p1":
        return "digest+p1" if producer == "embed" else "gist-hf+p1"
    return "digest"


def append_then_pool(
    prev_c: np.ndarray | None,
    new_rows: np.ndarray,
    k_max: int = DEFAULT_K_MAX,
    ema: float = DEFAULT_EMA,
    protect_mask: list[bool] | None = None,
    protect_kinds_tags: list[set[str]] | None = None,
) -> np.ndarray:
    """Append new gist rows, then EMA-merge highest-cosine adjacent pair until k <= k_max.

    When ``protect_mask`` is provided (aligned with the stacked rows after concat),
    pairs where either row is protected are ineligible for EMA merge. Overflow then
    evicts oldest unprotected rows. Identifier-tagged protected rows are last to go.
    When ``protect_mask`` is None, behavior matches Cursor upstream.
    """
    if new_rows.ndim == 1:
        new_rows = new_rows[None, :]
    if prev_c is None or prev_c.size == 0:
        stacked = np.asarray(new_rows, dtype=np.float32)
        n_prev = 0
    else:
        prev = np.asarray(prev_c, dtype=np.float32)
        stacked = np.vstack([prev, np.asarray(new_rows, dtype=np.float32)])
        n_prev = int(prev.shape[0])

    n = int(stacked.shape[0])
    if protect_mask is None:
        protected = [False] * n
        tags: list[set[str]] = [set() for _ in range(n)]
    else:
        if len(protect_mask) != n:
            # Allow mask for new rows only (common when prev tags unknown).
            if len(protect_mask) == n - n_prev and n_prev >= 0:
                protected = [False] * n_prev + [bool(x) for x in protect_mask]
            else:
                raise ValueError(
                    f"protect_mask length {len(protect_mask)} != stacked rows {n}"
                )
        else:
            protected = [bool(x) for x in protect_mask]
        if protect_kinds_tags is None:
            tags = [{"protected"} if p else set() for p in protected]
        elif len(protect_kinds_tags) == n:
            tags = [set(t) for t in protect_kinds_tags]
        elif len(protect_kinds_tags) == n - n_prev:
            tags = [set() for _ in range(n_prev)] + [set(t) for t in protect_kinds_tags]
        else:
            tags = [{"protected"} if p else set() for p in protected]

    while stacked.shape[0] > k_max:
        n = int(stacked.shape[0])
        best_i = -1
        best_sim = -2.0
        for i in range(n - 1):
            if protected[i] or protected[i + 1]:
                continue
            sim = _cosine_adjacent(stacked[i], stacked[i + 1])
            if sim > best_sim:
                best_sim = sim
                best_i = i

        if best_i >= 0:
            merged = ema * stacked[best_i] + (1.0 - ema) * stacked[best_i + 1]
            stacked = np.vstack(
                [stacked[:best_i], merged[None, :], stacked[best_i + 2 :]]
            )
            protected = protected[:best_i] + [False] + protected[best_i + 2 :]
            tags = tags[:best_i] + [set()] + tags[best_i + 2 :]
            continue

        # No eligible merge pair — evict oldest unprotected, then non-identifier protected.
        evict = _pick_eviction_index(protected, tags)
        if evict is None:
            log.warning("matrix_over_protected k=%d k_max=%d", n, k_max)
            evict = 0
        stacked = np.vstack([stacked[:evict], stacked[evict + 1 :]])
        protected = protected[:evict] + protected[evict + 1 :]
        tags = tags[:evict] + tags[evict + 1 :]

    return l2_normalize(stacked.astype(np.float32))


def _pick_eviction_index(
    protected: list[bool], tags: list[set[str]]
) -> int | None:
    """Prefer oldest unprotected; then oldest protected non-identifier; never prefer identifiers."""
    for i, p in enumerate(protected):
        if not p:
            return i
    for i, t in enumerate(tags):
        if "identifier" not in t:
            return i
    # All identifier — documented hard-cap stress: evict oldest identifier last resort.
    return 0 if protected else None


def live_mask(k: int, k_max: int | None = None) -> np.ndarray:
    if k_max is None or k_max == k:
        return np.ones((k,), dtype=np.float32)
    mask = np.zeros((k_max,), dtype=np.float32)
    mask[:k] = 1.0
    return mask
